File: permutationsII.java
import java.util.*;
public class Solution {
    public ArrayList<ArrayList<Integer>> permuteUnique(int[] num) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        HashSet<ArrayList<Integer>> uniqueResult = new HashSet<ArrayList<Integer>>();
        ArrayList<Integer> temp = new ArrayList<Integer>();
        if (num.length == 1) {
            temp.add(num[0]);
            result.add(temp);
            return result;
        }
        int[] sub = new int[num.length-1];
        for (int i=0; i < sub.length; i++) {
            sub[i] = num[i+1];
        }
        ArrayList<ArrayList<Integer>> v = permuteUnique(sub);
        for (int i=0; i < v.size(); ++i) {
            for (int j=0; j <= v.get(i).size(); j++) {
                temp = (ArrayList<Integer>)v.get(i).clone();
                temp.add(j, num[0]);
                uniqueResult.add(temp);
            }
        }
        for (ArrayList<Integer> a:uniqueResult) {
            result.add(a);
        }
        return result;
    }
}

File: permutationsII.cpp
#include "leetcode.h"
class Solution {
public:
    vector<vector<int> > permuteUnique(vector<int> &num) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<vector<int> > result;
        set<vector<int> > uniqueResult;
        vector<int> temp;
        if (num.size() == 1) {
            temp.push_back(num[0]);
            result.push_back(temp);
            return result;
        }
        vector<int> sub(num.begin()+1, num.end());
        vector<vector<int> > v = permuteUnique(sub);
        for (int i=0; i < v.size(); ++i) {
            for (int j=0; j <= v[i].size(); j++) {
                temp = v[i];
                temp.insert(temp.begin()+j, num[0]);
                uniqueResult.insert(temp);
            }
        }
        for (set<vector<int> >::iterator it=uniqueResult.begin(); it!=uniqueResult.end(); ++it) {
            result.push_back(*it);
        }
        return result;
    }
};
int main() {
    Solution s;
    vector<int> in;
    in.push_back(1);
    in.push_back(1);
    in.push_back(3);
    vector<vector<int> > v;
    v = s.permuteUnique(in);
    for (int i=0; i < v.size(); ++i) {
        for (int j=0; j < v[i].size(); j++) {
            cout << v[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}
